language-support/java/codegen/src/main/scala/com/digitalasset/daml/lf/codegen/CodeGenRunner.scala

```diff
@@ -109,6 +109,78 @@ object CodeGenRunner extends StrictLogging {
     logger.warn(s"Finish writing file '$outputFile'")
   }
 
+  /** Given the package prefixes specified per DAR and the module-prefixes specified in
+    * daml.yaml, produce the combined prefixes per package id.
+    */
+  def resolvePackagePrefixes(
+      pkgPrefixes: Map[PackageId, String],
+      modulePrefixes: Map[PackageReference, String],
+      interfaces: Seq[Interface],
+  ): Map[PackageId, String] = {
+    val metadata: Map[PackageReference.NameVersion, PackageId] = interfaces.view
+      .flatMap(iface =>
+        iface.metadata.iterator.map(metadata =>
+          PackageReference.NameVersion(metadata.name, metadata.version) -> iface.packageId
+        )
+      )
+      .toMap
+    def resolveRef(ref: PackageReference): PackageId = ref match {
+      case nameVersion: PackageReference.NameVersion =>
+        metadata.getOrElse(
+          nameVersion,
+          throw new IllegalArgumentException(
+            s"""No package $nameVersion found, available packages: ${metadata.keys.mkString(
+              ", "
+            )}"""
+          ),
+        )
+    }
+    val resolvedModulePrefixes: Map[PackageId, String] = modulePrefixes.view.map { case (k, v) =>
+      resolveRef(k) -> v.toLowerCase
+    }.toMap
+    val resolvedPrefixes: Map[PackageId, String] =
+      (pkgPrefixes.keySet union resolvedModulePrefixes.keySet).view.map { k =>
+        val prefix = (pkgPrefixes.get(k), resolvedModulePrefixes.get(k)) match {
+          case (None, None) =>
+            throw new RuntimeException(
+              "Internal error: key in pkgPrefixes and resolvedModulePrefixes could not be found in either of them"
+            )
+          case (Some(a), None) => a.stripSuffix(".")
+          case (None, Some(b)) => b.stripSuffix(".")
+          case (Some(a), Some(b)) => s"""${a.stripSuffix(".")}.${b.stripSuffix(".")}"""
+        }
+        k -> prefix
+      }.toMap
+    resolvedPrefixes
+  }
+
+  /** Verify that no two module names collide when the given
+    * prefixes are applied.
+    */
+  def detectModuleCollisions(
+      pkgPrefixes: Map[PackageId, String],
+      interfaces: Seq[Interface],
+  ): Unit = {
+    val allModules: List[(String, PackageId)] = interfaces.view.flatMap { interface =>
+      val modules = interface.typeDecls.keys.view.map(name => name.module).toSet
+      modules.map(m =>
+        pkgPrefixes
+          .get(interface.packageId)
+          .fold(m.toString)(prefix => s"$prefix.$m") -> interface.packageId
+      )
+    }.toList
```

Comment:
```suggestion
    val allModules: Seq[(String, PackageId)] =
      for {
        interface <- interfaces
        modules = interface.typeDecls.keySet.map(_.module)
        module <- modules
        maybePrefix = pkgPrefixes.get(interface.packageId)
        prefixedName = maybePrefix.fold(module.toString)(prefix => s"$prefix.$module")
      } yield prefixedName -> interface.packageId
```
* Doesn't go through a view but for the number of items we expect here it's probably fine.
* Uses `.keySet` instead of `.keys ... .toSet`.
* Breaks down the logic in a couple of lines, should be more readable

Mostly a matter of taste I believe, feel free to use it or not.

Reply:
I like it, thanks!